python/variable.cpp

```diff
@@ -128,80 +186,96 @@ void bind_init_0D(py::class_<Variable> &c) {
-      .def("__radd__", [](Variable &a, double &b) { return a + b; },
-           py::is_operator())
-      .def("__rsub__", [](Variable &a, double &b) { return b - a; },
-           py::is_operator())
-      .def("__rmul__", [](Variable &a, double &b) { return a * b; },
-           py::is_operator())
-      .def("__repr__", [](const Variable &self) { return to_string(self); });
-
-  // For some reason, pybind11 does not convert python lists to py::array,
-  // so we need to bind the lists manually.
-  // TODO: maybe there is a better way to do this?
-  bind_init_1D<int32_t>(variable);
-  bind_init_1D<double>(variable);
-  // This should be in the certain order
-  bind_init_0D(variable);
-  bind_init_0D<bool>(variable);
-  bind_init_0D<int64_t>(variable);
-  bind_init_0D<double>(variable);
-  bind_init_0D<py::object>(variable);
+
+  binder.append({bind_init_0D<DataArray>, bind::Priority::NativeScipp});
+  binder.append({bind_init_0D<Dataset>, bind::Priority::NativeScipp});
+  binder.append({bind_init_0D<std::string>, bind::Priority::NativeScipp});
+  binder.append({bind_init_0D<Eigen::Vector3d>, bind::Priority::NativeScipp});
+
+  binder.append(
+      {[](py::class_<Variable> &variable) {
+         variable.def(py::init<const VariableProxy &>())
+             .def(py::init(&makeVariableDefaultInit),
+                  py::arg("dims") = std::vector<Dim>{},
+                  py::arg("shape") = std::vector<scipp::index>{},
+                  py::arg("unit") = units::Unit(units::dimensionless),
+                  py::arg("dtype") = py::dtype::of<double>(),
+                  py::arg("variances").noconvert() = false)
+             .def(py::init(&doMakeVariable), py::arg("dims"),
+                  py::arg("values"), // py::array
+                  py::arg("variances") = std::nullopt,
+                  py::arg("unit") = units::Unit(units::dimensionless),
+                  py::arg("dtype") = py::none())
+             .def("rename_dims", &rename_dims<Variable>, py::arg("dims_dict"),
+                  "Rename dimensions.")
+             .def("copy", [](const Variable &self) { return self; },
+                  "Return a (deep) copy.")
+             .def("__copy__", [](Variable &self) { return Variable(self); })
+             .def("__deepcopy__",
+                  [](Variable &self, py::dict) { return Variable(self); })
+             .def_property_readonly("dtype", &Variable::dtype)
+             .def("__radd__", [](Variable &a, double &b) { return a + b; },
+                  py::is_operator())
+             .def("__rsub__", [](Variable &a, double &b) { return b - a; },
+                  py::is_operator())
+             .def("__rmul__", [](Variable &a, double &b) { return a * b; },
+                  py::is_operator())
+             .def("__repr__",
+                  [](const Variable &self) { return to_string(self); });
+       },
+       bind::Priority::NumpyArray});
+
+  binder.append({bind_init_1d_list, bind::Priority::PythonListNamedValues});
```

Comment:
Can we avoid (most or all?) priority names? Instead have something like

```cpp
binder.append<Dataset>(bind_init_0D<Dataset>);
binder.append<py::list>(bind_init_1d_list);
binder.append<bool>(bind_init_0D_native_python_types<bool>);
// ...
```
and translate type to `bind::Priority` internally?

If it works, I think it would be easier to remember. Would it also solve the `bool` and `int64_t` ordering issue?

Reply:
Not sure if it is possible, for instance, if you want to split 2 functions `f1(py::list, py::list)` and `f2(py::list, py::list)`: the first argument is the list of `Dim`, the second argument in `f1` is a `shape`, the second argument in `f2` is `values` and the convention is that if the second argument is just the positional one it should be treated like `shape`, the only way to provide such a behavior is to bind `f1` before `f2`, but the program doesn't know about this if you don't say it putting the priority manually.

Reply:
Sounds like a rare case?  I am not suggesting to *remove* the manual specification of the priority, but in most cases a type-based approach should be enough?

Reply:
Well, I can play with it, but if in this rare cases the user would rely on this kind of automation, very unclear bugs would appear. The additional question is what are the proper categories for the `enum`?

Reply:
> The additional question is what are the proper categories for the enum?
I don't know, I had assumed what you did captures everything we need? Is this not true?

Reply:
I've captured all we need right now, but I wouldn't say it is something very consistent. Or may be you could see the other categories I've missed.